Warnings from WorkGroup/Dispatcher.hpp:678

Hi @MrBurmark, could you take look at the following warnings? 

They come up on the following platform + compiler: toss3-gcc-8.3.1 

RAJA/include/RAJA/pattern/WorkGroup/Dispatcher.hpp:678:5: warning: operation on 'id' may be undefined [-Wsequence-point]

/RAJA/include/RAJA/pattern/WorkGroup/Dispatcher.hpp: In instantiation of 'static constexpr RAJA::detail::Dispatcher<platform, RAJA::policy::workgroup::direct_dispatch<T0, T1, TNs ...>, DispatcherID, CallArgs ...>::id_type RAJA::detail::Dispatcher<platform, RAJA::policy::workgroup::direct_dispatch<T0, T1, TNs ...>, DispatcherID, CallArgs ...>::get_id(camp::int_seq<int, id_types ...>, camp::list<Ts ...>) [with T = RAJA::detail::HoldForall<RAJA::policy::sequential::seq_exec, RAJA::TypedRangeSegment<long int>, testWorkGroupOrderedSingle<ExecPolicy, OrderPolicy, StoragePolicy, DispatchTyper, IndexType, Allocator, WORKING_RES>::operator()(IndexType, IndexType) const [with ExecPolicy = RAJA::policy::sequential::seq_work; OrderPolicy = RAJA::policy::workgroup::ordered; StoragePolicy = RAJA::policy::workgroup::array_of_pointers; DispatchTyper = detail::direct_dispatch_typer; IndexType = long int; Allocator = detail::ResourceAllocator<camp::resources::v1::Host>::std_allocator<char>; WORKING_RES = camp::resources::v1::Host]::<lambda(long int)>, long int>; int ...id_types = {0, 1}; Ts = {RAJA::detail::HoldForall<RAJA::policy::sequential::seq_exec, RAJA::TypedRangeSegment<long int, long int>, testWorkGroupOrderedSingle<ExecPolicy, OrderPolicy, StoragePolicy, DispatchTyper, IndexType, Allocator, WORKING_RES>::operator()(IndexType, IndexType) const [with ExecPolicy = RAJA::policy::sequential::seq_work; OrderPolicy = RAJA::policy::workgroup::ordered; StoragePolicy = RAJA::policy::workgroup::array_of_pointers; DispatchTyper = detail::direct_dispatch_typer; IndexType = long int; Allocator = detail::ResourceAllocator<camp::resources::v1::Host>::std_allocator<char>; WORKING_RES = camp::resources::v1::Host]::<lambda(long int)>, long int>, RAJA::detail::HoldForall<RAJA::policy::sequential::seq_exec, RAJA::TypedRangeSegment<long int, long int>, testWorkGroupOrderedSingle<ExecPolicy, OrderPolicy, StoragePolicy, DispatchTyper, IndexType, Allocator, WORKING_RES>::operator()(IndexType, IndexType) const [with ExecPolicy = RAJA::policy::sequential::seq_work; OrderPolicy = RAJA::policy::workgroup::ordered; StoragePolicy = RAJA::policy::workgroup::array_of_pointers; DispatchTyper = detail::direct_dispatch_typer; IndexType = long int; Allocator = detail::ResourceAllocator<camp::resources::v1::Host>::std_allocator<char>; WORKING_RES = camp::resources::v1::Host]::<lambda(long int)>, long int>}; T0 = RAJA::detail::HoldForall<RAJA::policy::sequential::seq_exec, RAJA::TypedRangeSegment<long int>, testWorkGroupOrderedSingle<ExecPolicy, OrderPolicy, StoragePolicy, DispatchTyper, IndexType, Allocator, WORKING_RES>::operator()(IndexType, IndexType) const [with ExecPolicy = RAJA::policy::sequential::seq_work; OrderPolicy = RAJA::policy::workgroup::ordered; StoragePolicy = RAJA::policy::workgroup::array_of_pointers; DispatchTyper = detail::direct_dispatch_typer; IndexType = long int; Allocator = detail::ResourceAllocator<camp::resources::v1::Host>::std_allocator<char>; WORKING_RES = camp::resources::v1::Host]::<lambda(long int)>, long int>; T1 = RAJA::detail::HoldForall<RAJA::policy::sequential::seq_exec, RAJA::TypedRangeSegment<long int>, testWorkGroupOrderedSingle<ExecPolicy, OrderPolicy, StoragePolicy, DispatchTyper, IndexType, Allocator, WORKING_RES>::operator()(IndexType, IndexType) const [with ExecPolicy = RAJA::policy::sequential::seq_work; OrderPolicy = RAJA::policy::workgroup::ordered; StoragePolicy = RAJA::policy::workgroup::array_of_pointers; DispatchTyper = detail::direct_dispatch_typer; IndexType = long int; Allocator = detail::ResourceAllocator<camp::resources::v1::Host>::std_allocator<char>; WORKING_RES = camp::resources::v1::Host]::<lambda(long int)>, long int>; TNs = {}; camp::resources::v1::Platform platform = (camp::resources::v1::Platform)1; DispatcherID = void; CallArgs = {camp::resources::v1::Host}; RAJA::detail::Dispatcher<platform, RAJA::policy::workgroup::direct_dispatch<T0, T1, TNs ...>, DispatcherID, CallArgs ...>::id_type = int]':

--- a/include/RAJA/index/ListSegment.hpp
+++ b/include/RAJA/index/ListSegment.hpp
@@ -121,6 +121,7 @@ public:
                    Index_type length,
                    camp::resources::Resource resource,
                    IndexOwnership owned = Owned)
+    : m_resource(nullptr), m_owned(Unowned), m_data(nullptr), m_size(0)
   {
     initIndexData(values, length, resource, owned);
   }

--- a/include/RAJA/pattern/WorkGroup/Dispatcher.hpp
+++ b/include/RAJA/pattern/WorkGroup/Dispatcher.hpp
@@ -675,7 +675,9 @@ struct Dispatcher<platform, ::RAJA::direct_dispatch<T0, T1, TNs...>,
   static constexpr id_type get_id(camp::int_seq<int, id_types...>, camp::list<Ts...>)
   {
     id_type id{-1};
-    camp::sink((std::is_same<T, Ts>::value ? ((id = id_types), 0) : 0)...);
+    using int_array = int[];
+    // quiet UB warning by sequencing assignment to id with list initialization
+    int_array {0, (std::is_same<T, Ts>::value ? ((id = id_types), 0) : 0)...};
     return id;
   }
 

--- a/include/RAJA/pattern/WorkGroup/Dispatcher.hpp
+++ b/include/RAJA/pattern/WorkGroup/Dispatcher.hpp
@@ -675,9 +675,9 @@ struct Dispatcher<platform, ::RAJA::direct_dispatch<T0, T1, TNs...>,
   static constexpr id_type get_id(camp::int_seq<int, id_types...>, camp::list<Ts...>)
   {
     id_type id{-1};
-    using int_array = int[];
     // quiet UB warning by sequencing assignment to id with list initialization
-    int_array {0, (std::is_same<T, Ts>::value ? ((id = id_types), 0) : 0)...};
+    int unused[] {0, (std::is_same<T, Ts>::value ? ((id = id_types), 0) : 0)...};
+    camp::sink(unused); // quiet unused var warning
     return id;
   }
 

